### ISSUE TITLE
Rule proposal: guard-clause

### ISSUE THREAD

**Issue by RDIL** [sindresorhus/eslint-plugin-unicorn#1862]
### Description

Guard clauses are a technique for writing cleaner code.  The idea is to avoid nesting logic in if-statements, and instead return early in methods.  That way, readability is enhanced due to the lack of indents.

### Fail

```js
function hello() {
    if (cond) {
        doLogic()
        doMoreLogic()
    }
}
```

### Pass

```js
function hello() {
    if (!cond) {
        return
    }

    doLogic()
    doMoreLogic()
}
```

### Additional Info

Previously proposed to ESLint: https://github.com/eslint/eslint/issues/10858

More reading materials:

- https://wiki.c2.com/?GuardClause
- https://en.wikipedia.org/wiki/Guard_(computer_science)
- https://refactoring.com/catalog/replaceNestedConditionalWithGuardClauses.html

**Comment by B2o5T** [comment]
Exist in Shopify https://github.com/Shopify/web-configs/blob/main/packages/eslint-plugin/docs/rules/prefer-early-return.md

**Comment by RDIL** [comment]
@B2o5T that's nice, but I think it would be beneficial to have it here, as Shopify's plugin includes many bloated dependencies like eslint-plugin-react for instance.

**Comment by philihp** [comment]
If you're gonna copy their rule, at least make it better in some way.

**Comment by RDIL** [comment]
No, I absolutely agree @philihp.  I think it could be improved e.g. by adding support for loops (early continue/break).

**Comment by fregante** [comment]
Duplicate of https://github.com/sindresorhus/eslint-plugin-unicorn/issues/16
